plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergBucketing.java

```diff
@@ -0,0 +1,219 @@
+import org.apache.iceberg.transforms.Transform;
+import org.apache.iceberg.transforms.Transforms;
+import org.apache.iceberg.types.Type;
+import org.apache.iceberg.types.Types;
+import org.apache.iceberg.types.Types.BooleanType;
+import org.apache.iceberg.types.Types.DecimalType;
+import org.apache.iceberg.types.Types.DoubleType;
+import org.apache.iceberg.types.Types.FloatType;
+import org.apache.iceberg.types.Types.UUIDType;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import java.math.BigDecimal;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.function.Function;
+
+import static io.trino.plugin.iceberg.PartitionTransforms.getBucketTransform;
+import static io.trino.plugin.iceberg.TypeConverter.toTrinoType;
+import static io.trino.spi.type.Decimals.isLongDecimal;
+import static io.trino.spi.type.Decimals.isShortDecimal;
+import static io.trino.spi.type.TypeUtils.writeNativeValue;
+import static java.lang.String.format;
+import static org.apache.iceberg.types.Type.TypeID.DECIMAL;
+import static org.assertj.core.api.AssertionsForClassTypes.assertThatThrownBy;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertNull;
+import static org.testng.Assert.assertTrue;
+
+public class TestIcebergBucketing
+{
+    private static final TypeManager TYPE_MANAGER = new TestingTypeManager();
+
+    @Test
+    public void testBucketNumberCompare()
+    {
+        // Only primitive types are covered here (logical types like timestamp, date or time map to long)
+        assertBucketAndHashEquals("string", "Trino rocks", 2131833594);
+
+        assertBucketAndHashEquals("int", null, null);
+        assertBucketAndHashEquals("int", 0, 1669671676);
+        assertBucketAndHashEquals("int", 300_000, 1798339266);
+        assertBucketAndHashEquals("int", Integer.MIN_VALUE, 74448856);
+        assertBucketAndHashEquals("int", Integer.MAX_VALUE, 1819228606);
+
+        assertBucketAndHashEquals("long", null, null);
+        assertBucketAndHashEquals("long", 0L, 1669671676);
+        assertBucketAndHashEquals("long", 300_000_000_000L, 371234369);
+        assertBucketAndHashEquals("long", Long.MIN_VALUE, 1366273829);
+        assertBucketAndHashEquals("long", Long.MAX_VALUE, 40977599);
+
+        assertBucketAndHashEquals("string", null, null);
+        assertBucketAndHashEquals("string", "", 0);
+        assertBucketAndHashEquals("string", "test string", 671244848);
+        assertBucketAndHashEquals("string", "\u5f3a\u5927\u7684Trino\u5f15\u64ce", 822296301); // 3-byte UTF-8 sequences (in Basic Plane, i.e. Plane 0)
+        assertBucketAndHashEquals("string", "\uD843\uDFFC\uD843\uDFFD\uD843\uDFFE\uD843\uDFFF", 775615312); // 4 code points: 20FFC - 20FFF. 4-byte UTF-8 sequences in Supplementary Plane 2
+
+        assertBucketAndHashEquals("decimal(1, 1)", null, null);
+        assertBucketAndHashEquals("decimal(1, 0)", "0", 1364076727);
```

Comment:
please make sure all test cases from  https://iceberg.apache.org/spec/#appendix-b-32-bit-hash-requirements are present here.

we should also cover "The types below are not currently valid for bucketing ..." types.
if current code throws for them -- that's fine. 
if it works -- that's fine for now too

Reply:
> please make sure all test cases from https://iceberg.apache.org/spec/#appendix-b-32-bit-hash-requirements are present here.

adding as a TODO comment